node/collation-generation/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+				true
+			}
+		}
+	}
+}
+
+impl<Context> Subsystem<Context> for CollationGenerationSubsystem
+where
+	Context: SubsystemContext<Message = CollationGenerationMessage>,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let subsystem = CollationGenerationSubsystem { config: None };
+
+		let future = Box::pin(subsystem.run(ctx));
+
+		SpawnedSubsystem {
+			name: "CollationGenerationSubsystem",
+			future,
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Subsystem(SubsystemError),
+	#[from]
+	OneshotRecv(oneshot::Canceled),
+	#[from]
+	Runtime(RuntimeApiError),
+	#[from]
+	Util(util::Error),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+async fn handle_new_activations<Context: SubsystemContext>(
+	config: Arc<CollationGenerationConfig>,
+	activated: &[Hash],
+	ctx: &mut Context,
+	sender: &mpsc::Sender<AllMessages>,
+) -> Result<()> {
+	// follow the procedure from the guide:
+	// https://w3f.github.io/parachain-implementers-guide/node/collators/collation-generation.html
+
+	for relay_parent in activated.iter().copied() {
+		let global_validation_data = Arc::new(
+			request_global_validation_data_ctx(relay_parent, ctx)
+				.await?
+				.await??,
+		);
+
+		let availability_cores = request_availability_cores_ctx(relay_parent, ctx)
+			.await?
+			.await??;
+
+		for core in availability_cores {
+			let (scheduled_core, assumption) = match core {
+				CoreState::Scheduled(scheduled_core) => {
+					(scheduled_core, OccupiedCoreAssumption::Free)
+				}
+				CoreState::Occupied(_occupied_core) => {
+					// TODO: https://github.com/paritytech/polkadot/issues/1573
+					continue;
+				}
+				_ => continue,
+			};
+
+			if scheduled_core.para_id != config.para_id {
+				continue;
+			}
+
+			// we get local validation data synchronously for each core instead of within the subtask loop,
+			// because we have only a single mutable handle to the context, so the work can't really be distributed
+			let local_validation_data = match request_local_validation_data_ctx(
+				relay_parent,
+				scheduled_core.para_id,
+				assumption,
+				ctx,
+			)
+			.await?
+			.await??
+			{
+				Some(local_validation_data) => local_validation_data,
+				None => continue,
+			};
+
+			let task_global_validation_data = global_validation_data.clone();
+			let task_config = config.clone();
+			let mut task_sender = sender.clone();
+			ctx.spawn("collation generation collation builder", Box::pin(async move {
+				let validation_data_hash =
+					validation_data_hash(&task_global_validation_data, &local_validation_data);
+
+				let collation = (task_config.collator)(&task_global_validation_data, &local_validation_data).await;
+
+				let pov_hash = collation.proof_of_validity.hash();
+
+				let ccr = CandidateReceipt {
+					commitments_hash: (CandidateCommitments {
+						upward_messages: collation.upward_messages,
+						head_data: collation.head_data,
+						..Default::default()
```

Comment:
Thanks for bringing this to my attention. I think the collation function should return all information necessary to generate the commitments.

And this part of the code should be generating the erasure-root by using `polkadot-erasure-coding`.

Reply:
ddbf4606 simplifies the protocol a bit, but it only works if the collator is able to access/generate all fields of `CandidateCommitments`. I believe that to be the case, but I'm not 100% sure.